mesh indexing failing with small scale values

RevoluPowered opened the pull request:
Fixes:
- #44655 fbx files not working in latest release
- previous behavior was culling scale values of 0.01 and det == 0.00004. which were valid in FBX land, det was not == 0, but in the cases of actual det==0 scenario which can be reproduced with node_3d_editor_plugin you can see it actually has a det == 0.
- node_3d_editor_plugin was creating gizmos with transforms with det == 0, this is safe to do now due to this change, I have also fixed this for the gizmos but the grid still needs this to be done for it. This relies on absolute zero values to change visibility so I fixed this.
- fixed fbx material mappings going to metallic property

I debugged this by putting checks on det == zero.
I checked bistro fbx and it imports fine, no dropped geometry anymore even with 0.01 scaling and det == 0.00004.

debug build
![image](https://user-images.githubusercontent.com/748770/103316994-1cf39100-4a22-11eb-8e1c-ce5854d430d4.png)

release_debug build (cpu time halves :D)
![image](https://user-images.githubusercontent.com/748770/103317214-c33f9680-4a22-11eb-8a53-f859a86aa6aa.png)

without this PR

![image](https://user-images.githubusercontent.com/748770/103345721-1698ff80-4a8a-11eb-9f84-c2763f6ab246.png)

akien-mga commented on editor/plugins/node_3d_editor_plugin.cpp:
```diff
@@ -3281,6 +3281,20 @@ void Node3DEditorViewport::update_transform_gizmo_view() {
 
 	xform.basis.scale(scale);
 
+	// now it's calculated check it's scale if its invalid then disable it.
```
1 out of 3, can do better :P

> Now it's calculated check **its** scale if **it's** invalid then disable it.

RevoluPowered replied:
corrected

akien-mga commented on servers/rendering/renderer_scene_cull.cpp:
```diff
@@ -1160,7 +1160,8 @@ void RendererSceneCull::_update_instance(Instance *p_instance) {
 	p_instance->transformed_aabb = new_aabb;
 
-	if (p_instance->scenario == nullptr || !p_instance->visible || Math::is_zero_approx(p_instance->transform.basis.determinant())) {
+	// note: is equal approx means 0.01 scale won't work so don't do that.
```
That sounds weird, `is_zero_approx` uses `CMP_EPSILON` which is `0.00001`, that's 1000 times smaller than `0.01`.

RevoluPowered replied:
Smallest value which is valid to render are these:

so these aren't precision errors
```cpp
Smallest rendered value is 0.000004
Smallest rendered value is 0.000004
Smallest rendered value is 0.000004
Smallest rendered value is 0.000004
Smallest rendered value is 0.000004
Smallest rendered value is 0.000004
Smallest rendered value is 0.000003
```

determinant is returning these, but they're valid.

float can do much much smaller numbers albeit with precision errors but the value must be dealt with otherwise the rendered scene breaks.

RevoluPowered replied:
corrected

RevoluPowered committed:
Fix zero scaling and material mappings being mapped to wrong fields

- fixes scale values of 0.0013 (det == 0.00004) not rendering, they should render even at small values, but not at zero like the editor grid plugin supplies zero exactly.
- fixes node_3d_editor_plugin visibility bug when scale is zero
- fix culling with small scaling values - which are still valid to be rendered like 0.00004

note: grid is still not fixed, it has det == 0 issues but this fixes one of them.

AndreaCatania commented:
This PR fixes this issue: https://github.com/godotengine/godot/issues/45067

akien-mga commented:
Needs a rebase @RevoluPowered.

reduz commented:
Current BVH does not really allow flat AABBs so this fix will do, but I want to replace by one that does soon so this fix will be irrelevant by then. I suggest lets just merge for now so things work.

RevoluPowered commented:
Rebase is completed.

akien-mga commented:
Thanks!
